File: data_structures/trees/binary_search_tree.js
class Node {
    constructor(value) {
        this.value = value;
        this.leftNode = null;
        this.rightNode = null;
    }

    height() {
        if (this.leftNode == null && this.rightNode == null) {
            return 0
        }
        const leftHeight = this.leftNode ? this.leftNode.height() : null
        const rightHeight = this.rightNode ? this.rightNode.height() : null
        if (leftHeight > rightHeight) {
            return leftHeight + 1
        } else {
            return rightHeight + 1
        }
    }
}

class Tree {
    constructor() {
        this.root = null;
    }

    recursiveInsert(node, newNode) {
        if (node.value < newNode.value) {
            if (node.rightNode == null) {
                return node.rightNode = newNode;
            }
            return this.recursiveInsert(node.rightNode, newNode)
        }
        else {
            if (node.leftNode == null) {
                return node.leftNode = newNode
            }
            return this.recursiveInsert(node.leftNode, newNode)
        }
    }

    insertValue(value) {
        const newNode = new Node(value)
        if (this.root == null) {
            return this.root = newNode
        }
        return this.recursiveInsert(this.root, newNode)
    }

    exist(value, node = this.root) {
        if (node == null) {
            return false
        }
        if (node.value == value) {
            return true
        }
        if (node.value < value) {
            return this.exist(value, node.rightNode)
        } else {
            return this.exist(value, node.leftNode)
        }
    }

    print(node = this.root) {
        if (node.leftNode != null) { this.print(node.leftNode) }
        console.log(node.value)
        if (node.rightNode != null) { this.print(node.rightNode) }
    }

    height() {
        return this.root.height()
    }

    reverseTree(node = this.root) {
        if (node.leftNode != null) { this.reverseTree(node.leftNode) }
        if (node.rightNode != null) { this.reverseTree(node.rightNode) }
        let nodeAux = node.leftNode
        node.leftNode = node.rightNode
        node.rightNode = nodeAux
    }

    findMinValue(node) {
        if (!node.leftNode) { return node.value }
        return this.findMinValue(node.leftNode)
    }

    findMaxValue(node) {
        if (!node.rightNode) { return node.value }
        return this.findMaxValue(node.rightNode)
    }

    removeNode(value, node) {
        if (node == null) {
            return node;
        }

        if (node.value < value) {
            node.rightNode = this.removeNode(value, node.rightNode)
        }
        else if (node.value > value) {
            node.leftNode = this.removeNode(value, node.leftNode)
        }

        else {

            if (!node.leftNode && !node.rightNode) {
                return node = null
            }
            if (!node.leftNode) {
                return node.rightNode
            }
            else if (!node.rightNode) {
                return node.leftNode
            }

            node.value = this.findMinValue(node.rightNode)
            node.rightNode = this.removeNode(node.value, node.rightNode)

        }
        return node
    }

    removeValue(value) {
        this.root = this.removeNode(value, this.root)
    }
}


const myTree = new Tree();
myTree.insertValue(9)
myTree.insertValue(13)
myTree.insertValue(7)
myTree.insertValue(5)
myTree.insertValue(2)
myTree.insertValue(1)
myTree.insertValue(15)
myTree.insertValue(21)
console.log("Tree")
myTree.print()
//           9
//         /   \
//        7     13
//       /       \
//       5       15
//      /         \
//     2          21
//    /
//   1

myTree.print()
console.log(`Min value = ${myTree.findMinValue(myTree.root)}`)
console.log(`Max value = ${myTree.findMaxValue(myTree.root)}`)

console.log("Tree after delete leaf node")
myTree.removeValue(21)
myTree.print()

console.log("Tree after delete root node")
myTree.removeValue(9)
myTree.print()

//myTree.reverseTree()
//console.log("Reversed Tree")
console.log(JSON.stringify(myTree))

File: data_structures/trees/binary_search_tree.py
class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None
        self.depth = 1
    
    def height(self):
        if self.left is None and self.right is None: return 0

        right_height = self.right.height() if self.right is not None else 0
        left_height = self.left.height() if self.left is not None else 0

        if right_height > left_height:
            return right_height + 1
        else:
            return left_height + 1


class Tree:
    def __init__(self):
        self.root = None

    def recursiveInsert(self, node, newNode):
        if node.value > newNode.value:
            if node.left is None:
                newNode.depth = node.depth +1
                node.left = newNode
                return newNode
            self.recursiveInsert(node.left, newNode)
        else:
            if node.right is None:
                newNode.depth = node.depth +1
                node.right = newNode
                return newNode
            self.recursiveInsert(node.right, newNode)

    
    def insert(self, value):
        new_node = Node(value)

        if self.root is None: 
            self.root = new_node
            return new_node
        
        return self.recursiveInsert(self.root, new_node)
    
    def printTree(self, node = None, position = None):
        if self.root == None: 
            print(None)
            return None
        if node is None: 
            node = self.root
            position = "root"

        print(position, end=": ")
        print(node.value)
        
        if node.left is not None: 
            if node.left.depth > node.depth:
                position = f"leafs of {node.value}"
            self.printTree(node.left, position= position + " left")
        if node.right is not None: 
            if node.right.depth > node.depth:
                position = f"leafs of {node.value}"
            self.printTree(node.right, position= position + " right")
    
    def height(self):
        if self.root is None: return 0
        return self.root.height()
    
    def depth(self, node = None):
        if self.root == None: return 0
        if node == None: node = self.root

        depth_right = node.depth if node.right is None else self.depth(node.right)
        depth_left = node.depth if node.left is None else self.depth(node.left)

        if depth_right > depth_left:
            return depth_right
        return depth_left
    
    def findMin(self, node = None):
        if self.root is None: return None
        if node == None: node = self.root

        return node.value if node.left is None else self.findMin(node.left)
    
    def findMax(self, node = None):
        if self.root is None: return None
        if node == None: node = self.root

        return node.value if node.right is None else self.findMax(node.right)

    def swap(self, node = None):
        if self.root is None: return None
        if node is None: node = self.root

        if node.left: self.swap(node.left)
        if node.right: self.swap(node.right)

        node.left, node.right = node.right, node.left

       

tree = Tree()
print("empty tree")
tree.printTree()
print(f"height: {tree.height()}")
print(f"depth: {tree.depth()}")
print(f"min: {tree.findMin()}")
print(f"max: {tree.findMax()}")
tree.insert(2)
tree.insert(5)
tree.insert(1)
tree.insert(4)
tree.insert(3)
tree.insert(9)
tree.insert(10)
print("tree after insertion")
tree.printTree()
print(f"height: {tree.height()}")
print(f"depth: {tree.depth()}")
print(f"min: {tree.findMin()}")
print(f"max: {tree.findMax()}")
tree.swap()
print("tree after swap")
tree.printTree()